File: compiler/src/queries/CommentQueries.swift
extension Queries {
    public struct Comments {
        public var nodes: [Node]
        public var comments: [String]
        public var links: [String: Link]
    }

    public static func comments(includeLinks: Bool) -> Query<Comments> {
        { db, node, body in
            let definitionNode = db[node, Resolved.self]?.definitions.first ?? node

            guard let definition = db[definitionNode, Defined.self]?.definition else { return }

            body(
                .init(
                    nodes: [definitionNode],
                    comments: definition.comments,
                    links: includeLinks
                        ? getLinks(db: db, definitionNode: definitionNode, sourceNode: node) : [:],
                )
            )
        }
    }

    public struct ErrorInstance {
        public let bound: Bound
        public let isDefault: Bool
        public let comments: Queries.Comments
        public let trace: [Constraint]
    }

    public static var errorInstance: Query<ErrorInstance> {
        { db, node, body in
            guard let bounds = db[node, Bounds.self]?.bounds else { return }

            for case (_, (let bound, let resolved?)) in bounds {
                guard resolved.instance.isError,
                    let instanceDefinition = db[resolved.instance.node, Defined.self]?.definition
                        as? InstanceDefinition
                else { continue }

                var comments = Comments(
                    nodes: [resolved.instance.node],
                    comments: instanceDefinition.comments,
                    links: getLinks(
                        db: db,
                        definitionNode: resolved.instance.node,
                        sourceNode: resolved.resolvedNode,
                    ),
                )

                let trace = bound.substitutions.values
                    .compactMap { type -> [Constraint]? in
                        guard case .node(let node) = type, let group = db[node, Typed.self]?.group
                        else { return nil }

                        comments.nodes.append(contentsOf: group.nodes)

                        return group.trace
                    }
                    .joined()

                body(
                    .init(
                        bound: bound,
                        isDefault: resolved.instance.isDefault,
                        comments: comments,
                        trace: Array(trace),
                    )
                )
            }
        }
    }
}

File: compiler/queries/src/comment_queries.rs
use crate::QueryCtx;
use wipple_core::{
    db::Node,
    render::Comments,
    typecheck::bounds::{ResolvedBound, ResolvedBounds},
    util::get_links,
    visit::{
        Resolved,
        definitions::{Defined, InstanceDefinition},
    },
};

pub fn comments(db: &QueryCtx<'_>, node: Node) -> Option<Comments> {
    let definition_node = db
        .get::<Resolved>(node)
        .and_then(|resolved| resolved.definitions.first().copied())
        .unwrap_or(node);

    let Defined(definition) = db.get(definition_node)?;

    Some(Comments {
        definition: definition_node,
        comments: definition.comments().to_vec(),
        links: get_links(db, definition_node, node),
    })
}

#[derive(Debug, Clone)]
pub struct ErrorInstance<'a> {
    pub bound: &'a ResolvedBound,
    pub is_default: bool,
    pub comments: Comments,
}

pub fn error_instances<'a>(db: &QueryCtx<'a>, node: Node) -> Vec<ErrorInstance<'a>> {
    let Some(ResolvedBounds(bounds)) = db.get(node) else {
        return Vec::new();
    };

    bounds
        .values()
        .filter_map(|result| {
            let bound = result.as_ref().ok()?;

            let instance = db
                .get(bound.instance.node)
                .and_then(|Defined(definition)| definition.downcast_ref::<InstanceDefinition>())?;

            if !instance.error {
                return None;
            }

            let Defined(definition) = db.get(bound.instance.node)?;
            let instance_definition = definition.downcast_ref::<InstanceDefinition>()?;

            let comments = Comments {
                definition: bound.instance.node,
                comments: instance_definition.comments.clone(),
                links: get_links(db, bound.instance.node, node),
            };

            Some(ErrorInstance {
                bound,
                is_default: instance.default,
                comments,
            })
        })
        .collect()
}
